keyboards/melody96/keymaps/zunger/keymap.c

```diff
@@ -0,0 +1,250 @@
+};
+
+// Layer bitfields.
+#define GREEK_LAYER (1UL << _GREEK)
+#define SHIFTGREEK_LAYER (1UL << _SHIFTGREEK)
+#define CADET_LAYER (1UL << _CADET)
+#define SHIFTCADET_LAYER (1UL << _SHIFTCADET)
+// The layers we don't touch.
+#define LAYER_MASK ~(GREEK_LAYER|SHIFTGREEK_LAYER|CADET_LAYER|SHIFTCADET_LAYER)
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  // We track these persistent globals and manage them on our own, rather than trying to rely on
+  // get_mods or the like, because this function is called *before* that's updated!
+  static bool shift_held = false;
+  static bool alt_held = false;
+  static bool greek_held = false;
+  static bool cadet_held = false;
+
+  // These are where we remember the values of lock states.
+  static bool shift_lock = false;
+  static uint32_t layer_lock = _QWERTY;
+
+  // Process any modifier key presses.
+  if (keycode == KC_LSHIFT || keycode == KC_RSHIFT) {
+    shift_held = record->event.pressed;
+  } else if (keycode == KC_LALT || keycode == KC_RALT) {
+    alt_held = record->event.pressed;
+  } else if (keycode == KC_GREEK) {
+    greek_held = record->event.pressed;
+  } else if (keycode == KC_CADET) {
+    cadet_held = record->event.pressed;
+  }
+
+  // Now let's transform these into the "cadet request" and "greek request."
+  const bool greek_request = (greek_held && !alt_held);
+  const bool cadet_request = (cadet_held || (greek_held && alt_held));
+
+  // Now, handle the lock keys. We store next_layer_lock in a local variable so that we can
+  // determine the layer to pick right now before we update layer_lock.
+  uint32_t next_layer_lock = layer_lock;
+  if (keycode == KC_CAPS) {
+    // If we're in QWERTY mode, caps lock is already going to be managed by the host OS, but by
+    // tracking it ourselves we can also usefully apply it to the GREEK and CADET layers.
+    if (record->event.pressed) {
+      shift_lock = !shift_lock;
+    }
+  } else if (keycode == KC_LAYER_LOCK) {
+    if (record->event.pressed) {
+      if (cadet_request) {
+        next_layer_lock = _CADET;
+      } else if (greek_request) {
+        next_layer_lock = _GREEK;
+      } else {
+        next_layer_lock = _QWERTY;
+      }
+    }
+  }
+
+  // OK! Now we know which buttons are being held, and the current and upcoming states of the locks.
+  // We can compute our new base layer. Remember that the CADET and GREEK keys act as their own
+  // antonyms if they match the layer lock -- e.g., if you have CADET locked, then CADET+X generates
+  // QWERTY-X.
+  uint32_t base_layer;
+  if (cadet_request) {
+    base_layer = (layer_lock == _CADET ? _QWERTY : _CADET);
+  } else if (greek_request) {
+    base_layer = (layer_lock == _GREEK ? _QWERTY : _GREEK);
+  } else {
+    base_layer = layer_lock;
+  }
+
+  const bool shifted = (shift_held != shift_lock);
+  uint32_t actual_layer;
```

Comment:
The layer mask isn't always 32 bit, and can be changed. 

If you could change to the typedef that we recently added? 

layer_state_t
```suggestion
  layer_state_t actual_layer;
```

Reply:
Updated as per above.